### PR TITLE
[GRPO] switch grpo liger loss to triton version

### PR DESCRIPTION
**Pull request by kashif**
# What does this PR do?

Fixes dapo loss type support, but needs to wait for upstream merge and release 

- [x] https://github.com/linkedin/Liger-Kernel/pull/939
- [x] https://github.com/linkedin/Liger-Kernel/pull/993
- [ ] liger-kernel release

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Swaps out the Liger-kernel loss implementation and its call/metric wiring, which can change training behavior and stability across loss types and modalities. Risk is moderated by expanded test coverage around Triton invocation and parameter updates.
> 
> **Overview**
> Switches `GRPOTrainer`’s `use_liger_kernel` path from the old `LigerFusedLinearGRPOLoss` (hidden-state + lm_head) implementation to Liger’s Triton-based `triton_grpo_loss`, computing logits directly and passing through additional loss controls (e.g. `delta`, SA/PO temperatures, bias-corrected KL, importance-sampling ratio).
> 
> Removes the `delta`/two-sided-loss restriction for Liger in `GRPOConfig`, drops the forward-redirection and `_get_last_hidden_state` plumbing, and adjusts metric extraction to match Triton’s returned metric layout.
> 
> Updates tests to patch/verify `triton_grpo_loss` calls (including `vllm_is_ratio` propagation), broadens VLM+Liger coverage across multiple `loss_type`s, and removes assertions tied to the previous fused-loss class.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 6dbb0208345467d781b3303eb052416a0811c106. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

**Comment by kashif** [comment]
@lewtun  would you be able to benchmark this branch for memory usage of grpo trainer?

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1748,45 +1689,58 @@ def _generate_and_score_completions(
...
+            "pixel_attention_mask",
+            "image_sizes",
+            "token_type_ids",
+        ]
+        for key in optional_keys:
+            if key in inputs:
+                model_inputs[key] = inputs[key]
+        model_inputs["use_cache"] = False
+        if "logits_to_keep" in self.model_kwarg_keys:
+            model_inputs["logits_to_keep"] = logits_to_keep + 1
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py:2086]: small suggestion since it wasn't clear to me why we're doing this:

```suggestion
            # Add 1 because the last logits of the sequence is later excluded
            model_inputs["logits_to_keep"] = logits_to_keep + 1
```

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1748,45 +1689,58 @@ def _generate_and_score_completions(
...
+
+        logits = model(**model_inputs).logits
+        logits = logits[:, -(logits_to_keep + 1) :, :].contiguous()
```
Comment (lewtun) [inline trl/trainer/grpo_trainer.py]: If `logits_to_keep` is in `model_kwarg_keys` we add 1 above to `model_inputs`, but then here we also add 1. Is that a form of double counting or is it correct as currently implemented?

**Review by lewtun — APPROVED** [review]
Thanks a lot for adding this @kashif ! LGTM with a question about adding 1 to `logits_to_keep` twice. Could you also add your nice plots showing the memory improvement with the Triton loss to the PR description?

Also, do we need to bump the minimum `liger` version?

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: checking wiht example... let me report back

Comment (qgallouedec) [reply trl/trainer/grpo_trainer.py]: did you find the answer?

**Comment by qgallouedec** [comment]
LGTM, it would be nice to have a quick comparison of the profiles, just to get an idea of the extent of the memory reduction.

**Comment by kashif** [comment]
comparison Red (old) and pink is triton based


W B Chart 01_12_2025, 18_30_08

File: trl/trainer/grpo_trainer.py (hunk 2 above)
Comment (kashif) [reply trl/trainer/grpo_trainer.py]: liger triton grpo loss expects `logits_to_keep + 1` logits

**Comment by kashif** [comment]
grpo_kernels_memory
grpo_kernels_comparison

**Comment by kashif** [comment]
@yukiu00 I have updated this PR as well as the corresponding Liger PR with your changes if you want to test

**Comment by yukiu00** [comment]
@kashif  Thanks! I'll give it a try.

**Comment by kashif** [comment]
@yukiu00  🙇🏽

**Comment by kashif** [comment]
@yukiu00 did you have any feedback on this PR?

File: trl/trainer/grpo_trainer.py [hunk 3]
```diff
@@ -546,11 +546,7 @@ def __init__(
             raise NotImplementedError(
                 "Liger Kernels don't currently support masking token positions based on entropy."
             )
-        if self.use_liger_kernel and not self.importance_sampling_level == "token":
-            raise NotImplementedError(
-                "Liger Kernels currently only support token-level importance sampling. Please set"
-                "`importance_sampling_level` to 'token'."
-            )
+        # Liger Kernels now support both "token" and "sequence" importance sampling levels
```
Comment (albertvillanova) [inline trl/trainer/grpo_trainer.py]: Since which version "sequence" is supported?

**Comment by yukiu00** [comment]
Thanks for the update @kashif!

I have a couple of questions/suggestions:

1. Memory improvement: The Triton version materializes logits (though limited by `logits_to_keep`), while the chunked version avoids this by fusing linear projection with loss computation in chunks. Intuitively, the chunked approach should be more memory-efficient, but the benchmarks show the opposite. Could you share some insight on why the Triton version achieves lower peak memory?

2. Missing loss types in tests: The Liger-Kernel side (PR linkedin/Liger-Kernel#993) now supports all 7 loss types including `luspo`, `cispo`, and `sapo`. The code here correctly passes `self.loss_type` through to `triton_grpo_loss`, so these should work. However, `test_training_vlm_and_liger` only parametrizes `["grpo", "bnpo", "dr_grpo", "dapo"]`. Should we add `luspo`, `cispo`, and `sapo` to the test coverage as well?

**Comment by kashif** [comment]
thanks! so the reason triton uses less memory is that the triton is chunking over the vocab size while the orginal chunking is over the batch,  and we were still materializing the logits when chunking over the batch see: https://github.com/linkedin/Liger-Kernel/blob/main/src/liger_kernel/chunked_loss/fused_linear_ppo.py#L364 

yes i will add the other loses, my bad.

**Comment by yukiu00** [comment]
Thanks for the clear explanation @kashif! The vocab-dimension chunking in Triton vs batch-dimension chunking makes a lot of sense. Nice work on the Triton migration. looking forward to seeing it merged!

**Comment by kashif** [comment]
changes have been merged in Liger main, so next waiting for a release

**Comment by qgallouedec** [comment]
closing as stale
